### ISSUE TITLE
Sparse eigsh works wrong on low-rank matrix

### ISSUE THREAD

**Issue by PavelStishenko** [cupy/cupy#5001]
* Conditions
CuPy Version : 9.0.0rc1
NumPy Version : 1.20.1
SciPy Version : 1.6.2
CUDA Build Version : 11010
CUDA Driver Version : 11020
CUDA Runtime Version : 11010

* Code to reproduce
```
import cupy as cp
from cupyx.scipy.sparse.linalg import eigsh
from cupy.testing import assert_allclose

cp.random.seed(123)
a = cp.random.random((5,5))
a = a @  a.T

w,v = cp.linalg.eigh(a)		# Etalon
print (w)

w_sparse,v = eigsh(a, 2)
print (w_sparse)
assert_allclose(w[-2:], w_sparse)	# OK


a_trunc = v[:,-2:] @ cp.diag(w[-2:]) @ v[:,-2:].T

w_trunc,_ = cp.linalg.eigh(a_trunc)
print (w_trunc)
assert_allclose(w[-2:], w_trunc[-2:])	# OK


w_trunc_sparse,_ = eigsh(a_trunc, 2)
print (w_trunc_sparse)
assert_allclose(w[-2:], w_trunc_sparse)		# Fails
```
* Error messages, stack traces, or logs
The last assert fails.
`eigsh` works fine with full-rank matrix, but gives completely wrong eigenvalues on low-rank (`a_trunc`) matrix.

This issue i similar to #5000 , but I think it is a different bug, just the same test case.

**Comment by toslunar** [comment]
The current implementation uses the Lanczos algorithm.

https://github.com/cupy/cupy/blob/c98064928c8242d0c6a07e2c714e6c811f684a4e/cupyx/scipy/sparse/linalg/_eigen.py#L133-L141
(`_lanczos_fast` has the same algorithm but it runs faster.)

Without additional treatments, the Lanczos algorithm suffers a precision issue when the chain of Krylov subspaces stops within `i_end` steps, (i.e. when `beta[i]` is small).  Low-rank matrices are such examples.  (For a matrix with rank r, Krylov subspace is at most (r+1)-dim.)
https://en.wikipedia.org/wiki/Lanczos_algorithm#Numerical_stability

**Comment by Mazay0** [comment]
@toslunar,, @anaruse, @takagi,  please, take a look at this PR draft: #5183
I have adopted approach from SciPy SVDS for detection of too small beta's. 
Сonstructive criticism is encouraged.

**Comment by kmaehashi** [comment]
https://github.com/cupy/cupy/pull/7356#issuecomment-1409859290

This issue should be fixed via #7356
